src/component/mongoose_component_connection.erl

```diff
@@ -0,0 +1,485 @@
+            {true, #jid{lserver = FromLServer}} ->
+                FromLServer =:= LServer;
+            %% If the admin does not want to check the from field when accept packets from any
+            %% address. In this case, the component can send packet of behalf of the server users.
+            _ ->
+                ok
+        end,
+    ToJid = jid:from_binary(exml_query:attr(El, <<"to">>, <<>>)),
+    IsStanza = (<<"iq">> =:= Name)
+        orelse (<<"message">> =:= Name)
+        orelse (<<"presence">> =:= Name),
+    case IsStanza andalso IsValidFromJid andalso (error =/= ToJid) of
+        true ->
+            Acc = element_to_origin_accum(StateData, FromJid, ToJid, NewEl),
+            ejabberd_router:route(FromJid, ToJid, Acc, NewEl);
+        false ->
+            ?LOG_INFO(#{what => comp_bad_request,
+                        text => <<"Not valid Name or error in FromJid or ToJid">>,
+                        stanza_name => Name, from_jid => FromJid, to_jid => ToJid}),
+            Err = jlib:make_error_reply(NewEl, mongoose_xmpp_errors:bad_request()),
+            send_xml(StateData, Err),
+            error
+    end,
+    keep_state_and_data.
+
+-spec try_register_routes(data(), retries()) -> fsm_res().
+try_register_routes(StateData, Retries) ->
+    case register_routes(StateData) of
+        {ok, Components} ->
+            send_xml(StateData, #xmlel{name = <<"handshake">>}),
+            {next_state, stream_established, StateData#component_data{components = Components}};
+        {error, Reason} ->
+            #component_data{listener_opts = #{conflict_behaviour := ConflictBehaviour}} = StateData,
+            RoutesInfo = lookup_routes(StateData),
+            ?LOG_ERROR(#{what => comp_registration_conflict,
+                         text => <<"Another connection from a component with the same name">>,
+                         component => StateData#component_data.lserver,
+                         reason => Reason, retries => Retries, routes_info => RoutesInfo,
+                         conflict_behaviour => ConflictBehaviour}),
+            handle_registration_conflict(ConflictBehaviour, RoutesInfo, StateData, Retries)
+    end.
+
+-spec register_routes(data()) -> any().
+register_routes(StateData) ->
+    #component_data{listener_opts = #{hidden_components := AreHidden}} = StateData,
+    Routes = get_routes(StateData),
+    Handler = mongoose_packet_handler:new(mongoose_component, #{pid => self()}),
+    mongoose_component:register_components(Routes, node(), Handler, AreHidden).
+
+-spec get_routes(data()) -> [jid:lserver()].
+get_routes(#component_data{lserver = Subdomain, is_subdomain = true}) ->
+    StaticDomains = mongoose_domain_api:get_all_static(),
+    DynamicDomains = mongoose_domain_api:get_all_dynamic(),
+    [ component_route(Subdomain, Domain) || {Domain, _} <- StaticDomains ]
+    ++ [ component_route(Subdomain, Domain) || {Domain, _} <- DynamicDomains ];
```

Comment:
I think that our intention when creating `mongoose_subdomain_core` and `mongoose_lazy_routing` was to avoid this... I think that in order to really support dynamic domains, this needs to work for any domains added dynamically after the registration. IMO we need to extend `mongoose_component` to work like `mongoose_subdomain_core`, where you can register patterns.

Reply:
Yeah, the thing is that components are dynamic in a different way than multi-tenancy is: components can connect and disconnect from MongooseIM with any domain (or subdomain) they want. They really can even make up very arbitrary domains, `google.com` if you want to, it doesn't even have to be a subdomain of an existing domain, routing for components happens before S2S+DNS and after local domains (see `router_modules` defaults). And when the component disconnects the domain is not served anymore (`mongoose_subdomain_core` would read and write to the database, which we don't want).

Reply:
Ok, idea reworked here: https://github.com/esl/MongooseIM/pull/4461 @chrzaszcz 🙂

Reply:
Nice, thanks :)